Implement a basic actor runtime, including queuing and switching

rjmccall opened the pull request:
Testing needs a lot of work.

rjmccall committed:
Extract executor stuff out into a separate header and
start introducing the idea of a swiftasync CC.

rjmccall committed:
[NFC] Add some conveniences for working with task priorities

rjmccall commented:
@swift-ci Please test

rjmccall committed:
Add a swift::atomic<T> which uses a better ABI than MSVC's 128-bit std::atomic.

rjmccall committed:
Add support for declaring thread-local variables in the runtime.

Use native thread-locals when available and simulated
thread-locals when not.  The simulation layer uses
pthread_getspecific.

Using TLS is significantly more annoying this way, but I kindof
like it because it reinforces that TLS accesses aren't as cheap
as they look.

rjmccall committed:
Add a basic default-actor implementation with support for
eagerly adopting and abandoning threads.

rjmccall commented:
@swift-ci Please test

ktoso reviewed (approved):
Spent a lot of time reviewing the enqueues and process funcs, overall looking good 👍 
I'm sure I missed something but we'll find this as we add tests...

Some minor questions but nothing blocking, very cool stuff 👏

rjmccall commented on stdlib/public/Concurrency/Actor.cpp:
```diff
@@ -0,0 +1,931 @@
+      MaxPriority = 8,
+      MaxPriority_width = JobFlags::Priority_width,
+
+      // FIXME: add a reference to the running thread ID so that we
+      // can boost priorities.
```
There's some redundancy here between the existence of jobs and the status, yeah.  And yeah, you can maybe use slightly different atomics if you have fixed buffer, although of course most of the expensive is in the barriers and contention rather than any specific complexity of the atomic instruction.

rjmccall commented on stdlib/public/Concurrency/Actor.cpp:
```diff
@@ -0,0 +1,931 @@
+/// The goal of this class is to encapsulate uses of the central variable.
+/// We want to potentially use a more efficient access pattern than
+/// ordinary thread-locals when that's available.
+class DefaultActorProcessingFrame {
```
I expect that we'll end up with TLS storing the current task and executor, and the latter can probably be made redundant with this, yeah.

ktoso replied:
Great, thank you for confirming 👍

rjmccall commented on stdlib/public/Concurrency/Actor.cpp:
```diff
@@ -0,0 +1,931 @@
+    // Try the compare-exchange, and try again if it fails.
+    if (!CurrentState.compare_exchange_weak(oldState, newState,
+          /*success*/ std::memory_order_release,
+          /*failure*/ std::memory_order_relaxed))
```
An enqueuing thread doesn't need to synchronize with the current state of the actor because it never actually looks at any of the memory there besides what it reads from the atomic itself.  It just puts the pointer it finds as the next pointer after the job it's enqueuing.  The sequencing of operations on the atomic is sufficient to order things.

Let me put that in a picture.  The simple enqueue/dequeue scenario looks like this:

```text
Thread #1    ...enqueue....\......................................
                            \ (release)
Actor atomic ................\..................\.................
                                                 \ (acquire)
Thread #2    .....................................\....dequeue....
```

Everything is properly ordered because of the release/acquire and the sequencing of operations on the atomic.

What you're suggesting is that we need to acquire before releasing in the enqueue, so that we have a linear ordering through the second enqueuing thread:

```text
Thread #1    ...enqueue....\.........................................................
                            \ (rel)
Actor atomic ................\.........\.................../.......\.................
                                        \ (acq)           / (rel)   \ (acq)
Thread #2    ............................\....enqueue..../....       \        .......
                                                                      \
Thread #3    ..........................................................\...dequeue...
```

This is clearly well-ordered; you can trace a nice line through both enqueuing threads to the dequeuing thread, and everything interchange on the atomic fully synchronizes.  But that full synchronization isn't actually important because the enqueuing threads never access any of the memory associated with other enqueued jobs, just the atomic object itself and the pointer it contains.  And reads from the atomic object are ordered after all preceding changes to it, not just the last one.  So this simpler diagram is also perfectly well synchronized:

```text
Thread #1    ...enqueue....\.........................................................
                            \ (rel)
Actor atomic ................\............................./.......\.................
                                                          / (rel)   \ (acq)
Thread #2    .................................enqueue..../....       \        .......
                                                                      \
Thread #3    ..........................................................\...dequeue...
```

(Of course there's still a relaxed load that we're not drawing here, and that does have to be ordered on the atomic after the last store, but the compare-exchange takes care of that: if it wasn't ordered on the atomic after the last store, the comparison would fail.)

And there's a very important reason that we can't access any of this indirect memory associated with the queue while enqueuing, which is that that memory is in fact only *temporarily* associated with the queue for the period while it's in the queue, and as soon as some thread dequeues it, it's not safe to access anymore.  Since an enqueue can never guarantee that that isn't happening concurrently, it can never touch any of that memory.  Since we can't touch it anyway, we don't need to acquire it.

ktoso replied:
I see! Thanks a lot for the analysis here, makes a lot of sense now.

rjmccall commented on stdlib/public/Concurrency/Actor.cpp:
```diff
@@ -0,0 +1,931 @@
+    } else {
+      curOverride->wakeAndCancel();
+    }
+  }
```
Sortof, yeah.  The general concept is that there's something we need to look at the next time we process the actor.  The specific use right now is that they're threads that want to take over processing the actor, but the concept could pretty easily accommodate other things.

rjmccall commented on stdlib/public/Concurrency/Actor.cpp:
```diff
@@ -0,0 +1,931 @@
+  Job *firstNewJob = nullptr;
+  Job *lastNewJob = nullptr;
+
+  while (first != firstAlreadyPreprocessed) {
+    // If the job is an override, add it to the list of override jobs
+    // that we need to wake up.  Note that the list of override jobs
+    // flows through NextJob; we must not use getNextJobInQueue because
+    // that touches queue-private state, and the override job is
+    // not enqueued on the actor, merely registered with it.
+    if (first.isOverride()) {
+      auto overrideJob = first.getAsOverride();
+      first = overrideJob->NextJob;
+      overrideJob->NextJob = JobRef::getPreprocessed(overridesToWake);
+      overridesToWake = overrideJob;
+      continue;
+    }
+
+    // If the job isn't an override, add it to the end of the list of
+    // jobs.  Note that this reverses the order of jobs; since enqueue()
+    // always adds jobs to the front, reversing the order effectively
+    // makes the actor queue FIFO, which is what we want.
+    // FIXME: but we should also sort by priority
+    auto newJob = first.getAsJob();
+    first = getNextJobInQueue(newJob);
+    if (!firstNewJob)
+      firstNewJob = newJob;
+    if (lastNewJob)
+      setNextJobInQueue(lastNewJob, JobRef::getPreprocessed(newJob));
+    lastNewJob = newJob;
+
+  } while (first.needsPreprocessing());
```
Heh, oops, thanks!

rjmccall commented on include/swift/ABI/Executor.h:
```diff
@@ -0,0 +1,119 @@
+
+/// An ExecutorRef isn't necessarily just a pointer to an executor
+/// object; it may have other bits set.
+class ExecutorRef {
```
I'm still blitzing forward with my original concept of an actor being an exclusive executor, but you're right, if we want to have multiple actors sharing an executor, then we should change some of the naming here.

ktoso replied:
Yup, all good with going forward with your initial design here. Just thinking if I got this right etc. So today executor is actor, but I keep thinking that it's "has-a"...

We can revisit once we have things up and running and a real use case to revisit this with (the NIO one I mention, I expect could be interesting for anything event loop based...), thanks!

rjmccall commented:
@swift-ci Please test

DougGregor commented:
@swift-ci please test

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test Windows

rjmccall commented:
@swift-ci Please smoke test

rjmccall committed:
Add four words to DefaultActor so that we have some room to grow.

96 bytes is a reasonable quantum for a naked actor, and a small
actor with actual storage can store a few things without going
past 128 bytes.

rjmccall commented:
@swift-ci Please test Windows

rjmccall committed:
Handle default actors by special-casing layout in IRGen instead
of adding a property.

This better matches what the actual implementation expects,
and it avoids some possibilities of weird mismatches.  However,
it also requires special-case initialization, destruction, and
dynamic-layout support, none of which I've added yet.

In order to get NSObject default actor subclasses to use Swift
refcounting (and thus avoid the need for the default actor runtime
to generally use ObjC refcounting), I've had to introduce a
SwiftNativeNSObject which we substitute as the superclass when
inheriting directly from NSObject.  This is something we could
do in all NSObject subclasses; for now, I'm just doing it in
actors, although it's all actors and not just default actors.
We are not yet taking advantage of our special knowledge of this
class anywhere except the reference-counting code.

I went around in circles exploring a number of alternatives for
doing this; at one point I basically had a completely parallel
"ForImplementation" superclass query.  That proved to be a lot
of added complexity and created more problems than it solved.
We also don't *really* get any benefit from this subclassing
because there still wouldn't be a consistent superclass for all
actors.  So instead it's very ad-hoc.

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test

rjmccall committed:
Disable the swift::atomic<..., 16> specialization until we can
figure out how to convince clang-cl to accept it.

We need to make progress.

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test
